From 289ee2e0ef6d623b6244b16dc95222f042c9fa48 Mon Sep 17 00:00:00 2001
From: Kory Maincent <kory.maincent@bootlin.com>
Date: Tue, 13 Jan 2026 18:08:50 +0000
Subject: [PATCH] board: stm32mp2: Fix fwu-mdata node update for NOR boot

The fwu-mdata devicetree node was incorrectly updated for NOR boot.
The previous implementation wrote the memory node path as a string
to the fwu-mdata-store property, but the FWU metadata driver expects
a phandle reference, causing the driver to fail loading the metadata.

Fix this by using a proper phandle ID instead of a string path.

Also add the mdata-parts property to specify the metadata partition
location within the NOR memory, as required by the FWU driver.

Signed-off-by: Kory Maincent <kory.maincent@bootlin.com>
---
 board/st/stm32mp2/stm32mp2.c | 43 +++++++++++++++++++++++++++++++++++-
 1 file changed, 42 insertions(+), 1 deletion(-)

diff --git a/board/st/stm32mp2/stm32mp2.c b/board/st/stm32mp2/stm32mp2.c
index 8ea962ef10a..17d7cffcdaf 100644
--- a/board/st/stm32mp2/stm32mp2.c
+++ b/board/st/stm32mp2/stm32mp2.c
@@ -812,12 +812,36 @@ void *env_sf_get_env_addr(void)
 #define HYPERFLASH_PATH "/soc@0/ommanager@40500000/memory-controller@40430000/flash@0"
 #endif
 
+static int fdt_update_fwu_mtdpart(void *blob, int nodeoff)
+{
+	int ret;
+
+	ret = fdt_increase_size(blob, 100);
+	if (ret) {
+		printf("fdt_increase_size: err=%s\n", fdt_strerror(ret));
+		return ret;
+	}
+
+	ret = fdt_setprop_string(blob, nodeoff, "mdata-parts", "metadata1");
+	if (ret) {
+		log_err("Can't set mdata-parts property to metadata1\n");
+		return ret;
+	}
+
+	ret = fdt_appendprop_string(blob, nodeoff, "mdata-parts", "metadata2");
+	if (ret)
+		log_err("Can't append metadata2 to mdata-parts property\n");
+
+	return ret;
+}
+
 int fdt_update_fwu_properties(void *blob, int nodeoff,
 			      const char *compat_str,
 			      const char *storage_path)
 {
 	int ret;
 	int storage_off;
+	u32 phandle;
 
 	ret = fdt_increase_size(blob, 100);
 	if (ret) {
@@ -831,14 +855,23 @@ int fdt_update_fwu_properties(void *blob, int nodeoff,
 		return ret;
 	}
 
+
 	storage_off = fdt_path_offset(blob, storage_path);
 	if (storage_off < 0) {
 		log_err("Can't find %s path\n", storage_path);
 		return nodeoff;
 	}
 
-	ret = fdt_setprop_string(blob, nodeoff, "fwu-mdata-store", storage_path);
+	phandle = fdt_get_phandle(blob, storage_off);
+	if (phandle < 0) {
+		log_err("Can't get phandle: err=%d\n", __func__, __LINE__, phandle);
+		return phandle;
+	} else if (phandle == 0) {
+		log_err("Can't get phandle: no phandle\n", __func__, __LINE__, phandle);
+		return -ENOENT;
+	}
 
+	ret = fdt_setprop_u32(blob, nodeoff, "fwu-mdata-store", phandle);
 	if (ret < 0)
 		log_err("Can't set fwu-mdata-store property\n");
 
@@ -874,11 +907,19 @@ int fdt_update_fwu_mdata(void *blob)
 		/* flash0 */
 		ret = fdt_update_fwu_properties(blob, nodeoff, "u-boot,fwu-mdata-mtd",
 						SPINAND_NOR_PATH);
+		if (ret)
+			return ret;
+
+		ret = fdt_update_fwu_mtdpart(blob, nodeoff);
 		break;
 	case BOOT_FLASH_HYPERFLASH:
 		/* flash0 */
 		ret = fdt_update_fwu_properties(blob, nodeoff, "u-boot,fwu-mdata-mtd",
 						HYPERFLASH_PATH);
+		if (ret)
+			return ret;
+
+		ret = fdt_update_fwu_mtdpart(blob, nodeoff);
 		break;
 	default:
 		/* TF-A firmware update not supported for other boot device */
-- 
2.43.0

